### PR TITLE
Pull Request for Issue1864: Add storage space monitor to dataman applications

### PR DESCRIPTION
**Pull request by iainworkman**
- Created a class, AMStorageInfo, which takes a lot of code from Qt5.4's QStorageInfo and provides space information on a given volume.
- Added a storage info class to AMDatamanAppController if local storage is being used.
- Created a timer mechanism in AMDatamanAppController which checks storage usage on a minute frequency when local storage is being used
- Added a progress bar to AMDatamanAppController to visualize storage usage when local storage is being used.
- Added an error mon call when the storage usage gets above 85% (suppressed after being shown once, so the user isn't spammed every minute with error mons).

### CODE REVIEW

**Comment by iainworkman** [comment]
@XuetaoLiu-CLS Would you mind taking a look at this one?
Thanks.

File: acquamanCommon.pri [hunk 1]
```diff
@@ -844,7 +845,8 @@ SOURCES += \
 	source/beamline/AMExclusiveStatesEnumeratedControl.cpp \
     source/ui/beamline/AMControlStopButton.cpp \
     source/ui/beamline/AMControlToolButton.cpp \
-    $$PWD/source/ui/AMToolButton.cpp
+    $$PWD/source/ui/AMToolButton.cpp \
```
Comment (XuetaoLiu-CLS) [inline acquamanCommon.pri]: just out of curious, why do we need `$$PWD/` for some files?

Comment (dretrex) [inline acquamanCommon.pri]: That is something from newer Qt Creators - maybe even only on Macs.

Comment (iainworkman) [inline acquamanCommon.pri]: I believe this is a factor of different Qt versions. 
When I add a class it appears in the project file in the format
```
source/blah/thing
```
whereas it seems as if when either Jessie or Darren add it it prepends the $$PWD onto the front. I have seen no observable difference between both. If the $$PWD prefix is removed everything still works fine.

File: source/application/AMDatamanAppController.cpp [hunk 2]
```diff
@@ -1875,3 +1932,33 @@ void AMDatamanAppController::onOpenOtherDatabaseClicked()
...
+
+		storageInfo_.refresh();
+		double storageUsed = double(storageInfo_.bytesTotal() - storageInfo_.bytesAvailable());
+		double percentageUsed = (storageUsed / storageInfo_.bytesTotal()) * 100;
+
+		if(percentageUsed > 85) {
+
+			if(!storageWarningProvided_) {
+				double percentageRemaining = 100 - percentageUsed;
+				AMErrorMon::alert(this, AMDATAMANAPPCONTROLLER_LOCAL_STORAGE_RUNNING_LOW, QString("Warning: Local storage space is at %1%. Please inform the beamline staff.").arg(percentageRemaining));
```
Comment (XuetaoLiu-CLS) [inline source/application/AMDatamanAppController.cpp]: Do we want a visual timeout message box for a visual alert?
```
	static void alert(const QObject *src = 0, int code = 0, const QString &desc = "", bool showTimeoutMsgBox=false, QString msgBoxTitle="Alert")
```

File: source/application/AMDatamanAppController.cpp [hunk 3]
```diff
@@ -1875,3 +1932,33 @@ void AMDatamanAppController::onOpenOtherDatabaseClicked()
...
+				storageWarningProvided_ = true;
```
Comment (XuetaoLiu-CLS) [inline source/application/AMDatamanAppController.cpp]: How will we clear the warning provided flag? By far, it seems that the only way to clear `storageWarningProvided_` is via:
 -- close acquaman
 -- remove/delete files on local storage to gain free space ? 10%
 -- restart acquaman

File: source/application/AMDatamanAppController.cpp (hunk 2 above)
Comment (iainworkman) [inline source/application/AMDatamanAppController.cpp]: That might be a good idea.

**Comment by XuetaoLiu-CLS** [comment]
Look great! Nice done :8ball: 

Time Estimate: 20 minutes

**Comment by dretrex** [comment]
Looks good to me too.  I'll build it on my mac first and if that works I'll wait for the comments from @XuetaoLiu-CLS to be addressed before merging.
Time Estimate: 5 minutes

**Comment by iainworkman** [comment]
Made those updates:
- The suppression check for the error mon is now based on an incremented integer. Instead of only displaying once the error mon is configured to only display every 30 checks where the space is found to be consecutively running low (30 minutes).
- Added a warning label which appears when space is running low.

**Comment by iainworkman** [comment]
@XuetaoLiu-CLS @dretrex You can have a look at this again now, if required.

**Comment by dretrex** [comment]
Also, I built it on my mac and it didn't build, but you've already addressed it with the ifdef with the includes.

**Comment by XuetaoLiu-CLS** [comment]
It looks good to me with the changes.

**Comment by dretrex** [comment]
I'm good too. I'll merge it in.  Good work everyone.

**Pull request merged**